quixstreams/core/stream/functions/apply.py

```diff
@@ -47,12 +47,23 @@ def wrapper(
                 key: Any,
                 timestamp: int,
                 headers: Any,
+                is_watermark: bool = False,
+                on_watermark=self.on_watermark,
             ) -> None:
                 # Execute a function on a single value and wrap results into a list
                 # to expand them downstream
-                result = func(value)
-                for item in result:
-                    child_executor(item, key, timestamp, headers)
+                if is_watermark:
+                    child_executor(
+                        value,
+                        key,
+                        timestamp,
+                        headers,
+                        True,
+                    )
+                else:
+                    result = func(value)
+                    for item in result:
+                        child_executor(item, key, timestamp, headers)
```

Comment:
That may not change the game entirely, but still.
```suggestion
                result = value if is_watermark else func(value)
                child_executor(result, key, timestamp, headers, is_watermark)
```
But with this `is_watermark` could also be left as a mandatory arg.